File: csharp/parameter_settings.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Drawing.Imaging;
using System.Linq;
using System.Threading;
using System.Windows.Forms;

using pcammls;
using static pcammls.pcammls;

namespace demo
{

	class Program
	{
		static void Main(string[] args)
        {
            Console.WriteLine("test start\n");

            PercipioSDK cl = new PercipioSDK();

            DeviceInfoVector dev_list = cl.ListDevice();
            int sz = dev_list.Count();
            if (sz == 0) {
                Console.WriteLine(string.Format("no device found."));
                return;
            }
			
			Console.WriteLine(string.Format("found follow devices:"));
            for (int idx = 0; idx < sz; idx++) {
                var item = dev_list[idx];
                Console.WriteLine("{0} -- {1} {2}", idx, item.id, item.modelName);
            }
            Console.WriteLine("select one:");
            int select = int.Parse(Console.ReadLine());

            System.IntPtr handle = cl.Open(dev_list[select].id);
            if (!cl.isValidHandle(handle))
            {
                Console.WriteLine(string.Format("can not open device!"));
                return;
            }

            int err = cl.DeviceStreamEnable(handle, PERCIPIO_STREAM_COLOR);
            if (err != TY_STATUS_OK)
            {
                Console.WriteLine(string.Format("device stream enable err:{0}", err));
                return;
            }
            //bool:color aec
            DevParam aec = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_BOOL_AUTO_EXPOSURE);
            if (aec.isEmpty())
                Console.WriteLine(string.Format("aec is not support!"));
            else { 
                Console.WriteLine(string.Format("current aec status : {0}", aec.toBool()));

                //disable color aec
                aec = cl.DevParamFromBool(false);
                err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_BOOL_AUTO_EXPOSURE, aec);
                Console.WriteLine(string.Format("aec close result : {0}", err));
            }

            // int:color exposure time
            DevParam exp = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_INT_EXPOSURE_TIME);
            if (exp.isEmpty())
                Console.WriteLine(string.Format("exposure time is not support!"));
            else { 
                Console.WriteLine(string.Format("current exposure time status : {0}, range : {1} - {2}, inc : {3}",
                    exp.toInt(), exp.mMin(), exp.mMax(), exp.mInc()));

                Console.WriteLine("Enter exposure time:");
                int exposure_time = int.Parse(Console.ReadLine());
                exp = cl.DevParamFromInt(exposure_time);
                err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_INT_EXPOSURE_TIME, exp);
                Console.WriteLine(string.Format("set color exposure time result : {0}", err));
            }

            //enum: color image mode
            DevParam image_mode = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_ENUM_IMAGE_MODE);
            if (image_mode.isEmpty())
                Console.WriteLine(string.Format("color image mode is not support!"));
            else {
                EnumEntryVector list = image_mode.eList();
                for (int idx = 0; idx < list.Count(); idx++) {
                    var mode = list[idx];
                    Console.WriteLine(string.Format("{0}: {1}x{2} - {3}", idx, cl.Width(mode), cl.Height(mode), cl.Description(mode)));
                }

                Console.WriteLine(string.Format("Enter image mode index:"));
                int index = int.Parse(Console.ReadLine());
                image_mode = cl.DevParamFromEnum(cl.Value(list[index]));
                err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_ENUM_IMAGE_MODE, image_mode);
                Console.WriteLine(string.Format("set color image mode result : {0}", err));
            }
            cl.DeviceStreamOn(handle);
            image_data img_parsed_color = new image_data();
             
            while (true) {
                FrameVector frames = cl.DeviceStreamRead(handle, 2000);
                for (int i = 0; i < frames.Count(); i++) {
                    image_data image = frames[i];
                    if (image.streamID == PERCIPIO_STREAM_COLOR)
                    {
                        cl.DeviceStreamImageDecode(image, img_parsed_color);
                        Console.WriteLine(string.Format("color image info : {0} x {1}", img_parsed_color.width, img_parsed_color.height));
                    }
                }
            }

            cl.DeviceStreamOff(handle);
            cl.Close(handle);

        }
    }
	
}

File: python/parameter_settings.py
'''
Description: 
Author: zxy
Date: 2023-07-14 09:48:00
LastEditors: zxy
LastEditTime: 2024-11-25 11:36:57
'''
import pcammls
from pcammls import * 
import cv2
import numpy
import sys
import os


def main():
    cl = PercipioSDK()

    dev_list = cl.ListDevice()
    for idx in range(len(dev_list)):
      dev = dev_list[idx]
      print ('{} -- {} \t {}'.format(idx,dev.id,dev.iface.id))
    if  len(dev_list)==0:
      print ('no device')
      return
    if len(dev_list) == 1:
        selected_idx = 0 
    else:
        selected_idx  = int(input('select a device:'))
    if selected_idx < 0 or selected_idx >= len(dev_list):
        return

    sn = dev_list[selected_idx].id
    
    handle = cl.Open(sn)
    if not cl.isValidHandle(handle):
      err = cl.TYGetLastErrorCodedescription()
      print('no device found : ', end='')
      print(err)
      return
    
    err = cl.DeviceStreamEnable(handle, PERCIPIO_STREAM_COLOR)
    if err:
       print('device stream enable err:{}'.format(err))
       return
    
    #bool:color aec
    aec = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_BOOL_AUTO_EXPOSURE)
    if aec.isEmpty():
        print('aec is not support!')
    else :
        print('current aec status : {}'.format(aec.toBool()))

        #disable color aec
        aec = cl.DevParamFromBool(False)
        err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_BOOL_AUTO_EXPOSURE, aec)
        print('aec close result : ', end='')
        print(err)

    #int:color exposure time
    exp = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_INT_EXPOSURE_TIME)
    if exp.isEmpty():
        print('exposure time is not support!')
    else :
        print('current exposure time status : {}, range : {} - {}, inc : {}'.format(exp.toInt(), exp.mMin(), exp.mMax(), exp.mInc()))

        exposure_time  = int(input('Enter exposure time:'))
        exp = cl.DevParamFromInt(exposure_time)
        err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_INT_EXPOSURE_TIME, exp)
        print('set color exposure time result : ', end='')
        print(err)

    image_mode = cl.DeviceGetParameter(handle, TY_COMPONENT_RGB_CAM, TY_ENUM_IMAGE_MODE)
    if image_mode.isEmpty():
        print('color image mode is not support!')
    else :
        list = image_mode.eList()
        for idx in range(len(list)):
            mode = list[idx]
            print('{}: {}x{} - {}'.format(idx, cl.Width(mode), cl.Height(mode), cl.Description(mode)))

        index = int(input('Enter image mode index:'))
        image_mode = cl.DevParamFromEnum(cl.Value(list[index]))
        err = cl.DeviceSetParameter(handle, TY_COMPONENT_RGB_CAM, TY_ENUM_IMAGE_MODE, image_mode)

    cl.DeviceStreamOn(handle)
    img_parsed_color        = image_data()
    while True:
        image_list = cl.DeviceStreamRead(handle, 2000)
        for i in range(len(image_list)):
            frame = image_list[i]
            if frame.streamID == PERCIPIO_STREAM_COLOR:
                img_color = frame
        
            cl.DeviceStreamImageDecode(img_color, img_parsed_color)
            mat_undistortion_color = img_parsed_color.as_nparray()
            cv2.imshow('rgb', mat_undistortion_color)

        k = cv2.waitKey(10)
        if k==ord('q'): 
          break

    cl.DeviceStreamOff(handle)    
    cl.Close(handle)
    pass

if __name__=='__main__':
    main()

